### PR TITLE
Rework the Iterating page

### PR DESCRIPTION
**Pull request by uzluisf**
* Discuss the Iterable and Iterator roles more thoroughly
* Develop the examples a bit more

### CODE REVIEW

**Comment by JJ** [comment]
In general, I'd like to see these extensive revisions responding to specific issues. It might seem an improvement, but if it's not addressing specific concerns, it might be just going in some direction that might be better, or worse. Besides, it needs revision, so I'd like PRs to be more compact, also if possible. I'll check it out anyway.

File: doc/Language/iterating.pod6 [hunk 1]
```diff
@@ -4,136 +4,321 @@
 
 =SUBTITLE Functionalities available for visiting all items in a complex data structure
 
-=head1 The C<Iterator> and C<Iterable> roles
+Similar to other many programming languages, Raku makes a fine distinction among
```
Comment (JJ) [inline doc/Language/iterating.pod6:7]: There's no good reason to eliminate a header. You can change its name, but it helps index the page and navigation.

File: doc/Language/iterating.pod6 [hunk 2]
```diff
@@ -4,136 +4,321 @@
...
+the terms I<iteration>, I<iterable>, and I<iterator>. Familiarizing yourself
```
Comment (JJ) [inline doc/Language/iterating.pod6:8]: iteration is nothing like iterable, Why would someone *not* want to make that distinction?

File: doc/Language/iterating.pod6 [hunk 3]
```diff
@@ -4,136 +4,321 @@
...
+=item An B<iteration> is a mechanism by which a block of code is executed repeatedly.
 
-The C<Iterable> role is relatively simple. It provides a stub for the
-C<iterator> method, which is the one actually used by statements such as C<for>.
-C<for> will call C<.iterator> on the variable it precedes, and then run a block
-once for every item. Other methods, such as array assignment, will make the
-C<Iterable> class behave in the same way.
+=item An B<iterable> is an object that produce an iterator.
+
+=item An B<iterator> is an object that performs an iteration. In other words, an
```
Comment (JJ) [inline doc/Language/iterating.pod6:16]: if "iteration" is "a block of code", "perform an iteration" can't be "compute the next value". Different things.

File: doc/Language/iterating.pod6 [hunk 4]
```diff
@@ -4,136 +4,321 @@
...
+iterator knows how to compute the next value during an iteration.
+
+Thus, an B<iterator> contains the state that moves as it works through the
```
Comment (JJ) [inline doc/Language/iterating.pod6:19]: "moves" or "changes"? Does it "hold" the value of the state or "contains" the state? Arguably, the state would be contained in iterable.

File: doc/Language/iterating.pod6 [hunk 5]
```diff
@@ -4,136 +4,321 @@
...
+available values. On the contrary, an B<iterable> is the originating container
+that has no need to contain the state corresponding to the process of iteration.
```
Comment (JJ) [inline doc/Language/iterating.pod6:21]: This is quite arguable...

File: doc/Language/iterating.pod6 [hunk 6]
```diff
@@ -4,136 +4,321 @@
...
+
+Raku provides support for both iterables and iterators via the
+L<C<Iterable>|/type/Iterable> and L<C<Iterator>|/type/Iterator> roles.
+
+=head1 The C<Iterable> role
+
+The L<C<Iterable>|/type/Iterable> role is relatively simple and it requires that
+the composed class provides an C<iterator> method, which must return
+an L<C<Iterator>|/type/Iterator> object. The L<C<iterator>|/type/Iterable#method_iterator>
+method is the construct actually used by statements such as C<for>, which
```
Comment (JJ) [inline doc/Language/iterating.pod6:31]: invoked probably better than used

File: doc/Language/iterating.pod6 [hunk 7]
```diff
@@ -4,136 +4,321 @@
...
+invokes C<iterator> on the variable it precedes, and then run a block once for
+every item. Similarly other statements, such as array assignment, will make the
+C<Iterable>-composed class behave in the same way.
+
+Any object that can return an C<Iterator> is an iterable. Most container types
```
Comment (JJ) [inline doc/Language/iterating.pod6:36]: Even if it does not do that role?

File: doc/Language/iterating.pod6 [hunk 8]
```diff
@@ -4,136 +4,321 @@
...
-                       $chain ~~ /^^ <[ACGT]>+ $$ / and
-                       $chain.chars %% 3 } ) {
-        self.bless( :$chain );
+    
+    method new(Str $chain where { .match(/^^ <[ACGT]>+ $$/) && .chars %% 3 }) {
+        self.bless(:$chain);
+    }
+
+    method iterator(DNA:D:) { 
+        $.chain.split('', :skip-empty).rotor(3).iterator
```
Comment (JJ) [inline doc/Language/iterating.pod6:63]: Why is this better than comb?

File: doc/Language/iterating.pod6 [hunk 9]
```diff
@@ -4,136 +4,321 @@
...
-my @longer-chain =  DNA.new('ACGTACGTT');
+.join.say for dna;
+# OUTPUT: «ACG␤TAC␤GTT␤ACT»
+
+my @longer-chain = DNA.new('ACGTACGTTACT');
 say @longer-chain.raku;
-# OUTPUT: «[("A", "C", "G"), ("T", "A", "C"), ("G", "T", "T")]␤»
+# OUTPUT: «[("A", "C", "G"), ("T", "A", "C"), ("G", "T", "T"), ("A", "C", "T")]␤»
+=end code
+
```
Comment (JJ) [inline doc/Language/iterating.pod6:89]: Mentions to "for" and "array assignment" are now further away from the example, so it might be worth the while to explain that in both case "iterator" is being called.

File: doc/Language/iterating.pod6 [hunk 10]
```diff
@@ -4,136 +4,321 @@
...
+The L<C<Iterator>|/type/Iterator> role is a bit more complex than C<Iterable>.
+An C<Iterator> object must have at least a L<C<pull-one>|/type/Iterator#method_pull-one>
+method that returns either a sentinel value
+L<C<IterationEnd>|/type/Iterator#index-entry-IterationEnd> or the value
+extracted from the collection during the iteration. 
+
+In addition to C<pull-one>, the C<Iterator> role provides a
+L<series of methods|/type/Iterator#Methods>, which allows for a finer operation
+of iteration in several contexts: adding or eliminating items, or skipping over
```
Comment (JJ) [inline doc/Language/iterating.pod6:100]: *the* iteration *mechanism*

File: doc/Language/iterating.pod6 [hunk 11]
```diff
@@ -4,136 +4,321 @@
...
+for all the other methods, so the only one that has to be defined is precisely
+C<pull-one>, of which only a stub is provided by the role. 
+
+While C<Iterable> provides the high-level interface loops will be working with,
+C<Iterator> provides the lower-level functions that will be called in every
+iteration of the loop. 
+
+Let's extend the C<DNA> class with this role. The demands for our iterable
+C<DNA> class are simple enough, however to motivate the discussion of the C<Iterator>
+role, we'll step it up a notch. When iterating over a C<DNA> instance object, we
```
Comment (JJ) [inline doc/Language/iterating.pod6:111]: colon here.

File: doc/Language/iterating.pod6 [hunk 12]
```diff
@@ -4,136 +4,321 @@
...
+want the reverse of any codon whose index is divisible by 2, and a rotated version
+of any codon whose index is divisible by 3. Whenever the codon's index is both
+divisible by 2 and 3, the case for divisible by 2 takes precedence. Otherwise,
+return the codon as-is.
+
```
Comment (JJ) [inline doc/Language/iterating.pod6:116]: Why would you want to do that?

File: doc/Language/iterating.pod6 [hunk 13]
```diff
@@ -4,136 +4,321 @@
...
 
-    method iterator( ){ self }
-    method pull-one( --> Mu){
-        if $!index < $.chain.chars {
-            my $codon = $.chain.comb.rotor(3)[$!index div 3];
-            $!index += 3;
-            return $codon;
-        } else {
-            return IterationEnd;
+    method iterator(DNA:D:) { self }
```
Comment (JJ) [inline doc/Language/iterating.pod6:192]: Good one, this one.

File: doc/Language/iterating.pod6 [hunk 14]
```diff
@@ -4,136 +4,321 @@
...
+    method pull-one(--> Mu) {
```
Comment (JJ) [inline doc/Language/iterating.pod6:193]: But then you need to do the same here.

File: doc/Language/iterating.pod6 [hunk 15]
```diff
@@ -4,136 +4,321 @@
...
+=head1 Composing both the C<Iterable> and C<Iterator> roles
+
+Both the C<Iterable> and C<Iterator> roles can be composed into a single
+class, and their required methods implemented:
 
 =begin code
 class DNA does Iterable does Iterator {
     has $.chain;
-    has Int $!index = 0;
+    has $!index = 0;
```
Comment (JJ) [inline doc/Language/iterating.pod6:186]: Why did you remove the Int?

File: doc/Language/iterating.pod6 [hunk 16]
```diff
@@ -4,136 +4,321 @@
...
+
+This need not always be the case, and in most cases C<iterator> will have to
 build an iterator type to be returned (that will, for instance, keep track of
 the iteration state, which we are doing now in the main class), such as we did
-in the previous example; however, this example shows the minimal code needed to
-build a class that fulfills the iterator and iterable roles.
+in the previous section; however, this example shows the minimal code needed to
+build a class that fulfills both the iterator and iterable roles.
+
+=head1 Manual iteration
```
Comment (JJ) [inline doc/Language/iterating.pod6:243]: If this is a subsection of the previous one, it should be done that way...

File: doc/Language/iterating.pod6 [hunk 17]
```diff
@@ -4,136 +4,321 @@
...
-C<pull-one> method. This one is going to be called every time a new iteration
-occurs, so it must keep the state of the last one. An C<$.index> attribute will
-hold that state across invocations; C<pull-one> will check if the end of the
-chain has been reached and will return the C<IterationEnd> constant provided by
-the role. Implementing this low-level interface, in fact, simplifies the
-implementation of the C<Iterable> interface. Now the iterator will be the object
-itself, since we can call C<pull-one> on it to access every member in turn;
-C<.iterator> will thus return just C<self>; this is possible since the object
-will be, at the same time, C<Iterable> and C<Iterator>.
-
```
Comment (JJ) [inline doc/Language/iterating.pod6:99]: Don't think it's a good idea to just delete content where there's no issue. If there's an error here, point it out. If there's no error, try to repurpose it or extend it so that it fits better your narrative.

File: doc/Language/iterating.pod6 [hunk 18]
```diff
@@ -4,136 +4,321 @@
...
+my @languages = <perl python haskell ruby>;
+my $iter = @languages.iterator;
+
+say $iter.pull-one; #=> «perl␤» 
+say $iter.pull-one; #=> «python␤» 
+say $iter.pull-one; #=> «haskell␤» 
+say $iter.pull-one; #=> «ruby␤» 
+say $iter.pull-one; #=> «IterationEnd␤» 
+=end code
+
```
Comment (JJ) [inline doc/Language/iterating.pod6:265]: In the manuals we should always reflect intent (#1748 ). This can be easily figured out from the syntax, and it's not an actual use case for pull-one (which will probably be used only internally). Not clear it fits within the title of the page, also.

File: doc/Language/iterating.pod6 [hunk 19]
```diff
@@ -4,136 +4,321 @@
...
+After all the iterator's values are consumed, the returned value is the constant
+C<IterationEnd>. Once C<IterationEnd> has been generated, it's forbidden to make
+attempts to fetch more data, and behavior for doing so is undefined. The only
+valid use of the sentinel value C<IterationEnd> in a program is identity
```
Comment (JJ) [inline doc/Language/iterating.pod6:269]: I think that's explained in its own page. Not clear there's an added value to mention it here.

File: doc/Language/iterating.pod6 [hunk 20]
```diff
@@ -4,136 +4,321 @@
...
+=end code
+
+Since C<IterationEnd> is a constant, it's important that the variable we compare
+against it be bound, and not assigned. Hence, the use of the sigilless variable
+C<value> in the previous example.
 
 =head1 How to iterate: contextualizing and topic variables
 
-C<for> and other loops place the item produced in every iteration into the
+Loops, such as C<for>, place the item produced in every iteration into the
```
Comment (JJ) [inline doc/Language/iterating.pod6:292]: "Loops" don't do that, so this turn of the sentence is not totally correct. For and others do. "loop" doesn't place the item in the topic variable, or while IARC.

File: doc/Language/iterating.pod6 [hunk 21]
```diff
@@ -4,136 +4,321 @@
...
 L<topic variable C<$_>|/language/variables#index-entry-topic_variable>, or
 capture them into the variables that are declared along with the block. These
 variables can be directly used inside the loop, without needing to declare them,
 by using the
 L<C<^> twigil|/syntax/$CIRCUMFLEX_ACCENT#(Traps_to_avoid)_twigil_^>.
 
-Implicit iteration occurs when using the L<sequence operator|/language/operators#index-entry-..._operators>.
+Implicit iteration occurs when using the L<sequence operator|/language/operators#index-entry-..._operators>:
```
Comment (JJ) [inline doc/Language/iterating.pod6:299]: This is the kind of change that you should probably do directly.

File: doc/Language/iterating.pod6 [hunk 22]
```diff
@@ -4,136 +4,321 @@
...
+iterating construct that instead of running in sink context, returns an item
+every iteration. This
 L<Advent Calendar tutorial|https://perl6advent.wordpress.com/2009/12/23/day-23-lazy-fruits-from-the-gather-of-eden/>
-explains use cases for this kind of loops; in fact, gather is not so much a
+explains use cases for this kind of loops; in fact, C<gather> is not so much a
 looping construct, but a statement prefix that collects the items produced by
 C<take> and creates a list out of them.
 
-=head1 C<Classic> loops and why we do not like them
+=head1 Classical loops and why we do not like them
```
Comment (JJ) [inline doc/Language/iterating.pod6:316]: I think "Classic" is better here. "Classic-style" if you preffer, or in italics instead of monospace.

**Review by JJ — COMMENTED** [review]
Please check out my comments here. Thanks for the work, I appreciate it.

**Comment by uzluisf** [comment]
Thanks for taking time to go over this PR, JJ! I figured that given the number of issues with it is better to simply close it. I'll move it to a [Github gist](https://gist.github.com/uzluisf/6a045c5031c9f8d6ef7cb53f15366f36), and once I've the time try to address all the concerns you've highlighted here. I'll ping you on irc once I've addressed enough of them.

>It might seem an improvement, but if it's not addressing specific concerns, it might be just going in some direction that might be better, or worse.

The main issue I have with this page is that it makes many assumptions, which is fine if you know the ins and outs of the language. Things such as composing both `Iterator` and `Iterable` into a class, should I always compose both of them, what circumstances merit composing both of them vs just composing `Iterable` and providing the `Iterator` via another class, etc.

**Comment by JJ** [comment]
Thanks for understanding this. I appreciate the work you've done, anyway, and it's piecewise salvageable. Keep them coming!
